### PR TITLE
[ivalue] Add tuple constructor + to<std::tuple<Args...>>

### PR DESCRIPTION
**Pull request by bwasti**
This PR adds the ability to use tuples directly with the IValue constructor rather than the vector<IValue> approach

### CODE REVIEW

File: aten/src/ATen/core/ivalue_inl.h [hunk 1]
```diff
@@ -522,6 +522,25 @@ c10::optional<T> generic_to(
...
+  return std::make_tuple(
+      t[I].to<typename std::tuple_element<I, Tuple>::type>()...);
+}
+
+template <
+    typename... Args,
+    typename Indices = c10::guts::make_index_sequence<sizeof...(Args)>,
+    c10::guts::enable_if_t<
+        !c10::guts::disjunction<std::is_lvalue_reference<Args>...>::value,
+        std::nullptr_t> = nullptr>
```
Comment (ezyang) [inline aten/src/ATen/core/ivalue_inl.h]: It's probably better if we use SFINAE based on `is_constructible<IValue, Args>...`, if it works.

File: aten/src/ATen/core/ivalue.h [hunk 2]
```diff
@@ -173,6 +174,17 @@ struct CAFFE2_API IValue final {
 
   // Tuple
   IValue(c10::intrusive_ptr<ivalue::Tuple> v);
+
+  template <typename Tuple, std::size_t... I>
+  IValue(const Tuple& t, c10::guts::index_sequence<I...>);
```
Comment (ezyang) [inline aten/src/ATen/core/ivalue.h]: Can we please make this constructor private?

Comment (ezyang) [reply aten/src/ATen/core/ivalue.h]: Also, we seem to have a polyfill for `std::apply` in `C++17.h`, can we just use that directly here?

**Review by ezyang — CHANGES_REQUESTED** [review]
I think this could be simpler.

**Comment by ezyang** [comment]
cc'ed @smessmer for C++ guidance, and @zdevito to sanity check that IValue should take tuple arguments (I think it should)

File: aten/src/ATen/core/ivalue_inl.h [hunk 3]
```diff
@@ -609,6 +634,17 @@ inline IValue::IValue(c10::intrusive_ptr<ivalue::Tuple> v)
...
+template <
+    typename... Args,
+    c10::guts::enable_if_t<
+        !c10::guts::disjunction<
+            std::is_lvalue_reference<Args>...,
+            c10::guts::negation<std::is_constructible<IValue, Args>>...>::value,
+        std::nullptr_t>>
+inline IValue::IValue(const std::tuple<Args...>& t)
+    : IValue(
+          std::move(c10::guts::apply(c10::ivalue::Tuple::create<Args...>, t))) {
```
Comment (ezyang) [inline aten/src/ATen/core/ivalue_inl.h]: This is pleasing to my eye!

File: aten/src/ATen/core/ivalue_inl.h [hunk 4]
```diff
@@ -522,6 +528,25 @@ c10::optional<T> generic_to(
   return std::move(ivalue).to<T>();
 }
 
+template <typename Tuple, std::size_t... I>
+Tuple generic_to_tuple_impl(
+    const std::vector<IValue>& t,
+    c10::guts::index_sequence<I...>) {
+  return std::make_tuple(
+      t[I].to<typename std::tuple_element<I, Tuple>::type>()...);
+}
```
Comment (ezyang) [inline aten/src/ATen/core/ivalue_inl.h]: Nit: avoid dumping helper functions like this in global namespace; put it in `namespace detail { ... }` or similar

File: aten/src/ATen/core/ivalue_inl.h [hunk 5]
```diff
@@ -522,6 +528,25 @@ c10::optional<T> generic_to(
...
+
+template <
+    typename... Args,
+    typename Indices = c10::guts::make_index_sequence<sizeof...(Args)>,
+    c10::guts::enable_if_t<
+        !c10::guts::disjunction<std::is_lvalue_reference<Args>...>::value,
+        std::nullptr_t> = nullptr>
+std::tuple<Args...> generic_to(IValue ivalue, _fake_type<std::tuple<Args...>>) {
+  auto vals = ivalue.toTuple()->elements();
```
Comment (ezyang) [inline aten/src/ATen/core/ivalue_inl.h]: There should be a runtime test that the number of elements in the tuple matches the number of arguments in the parameter pack. I don't see that test in this function, or the function above.

**Review by ezyang — CHANGES_REQUESTED** [review]
Looks good, but we need that runtime test that the number of tuple elements matches the IValue...

File: aten/src/ATen/core/ivalue.cpp [hunk 6]
```diff
@@ -204,6 +204,9 @@ void ivalue::Object::resizeObject(size_t slot) {
   slots_.resize(type()->numAttributes());
 }
 
+
+ivalue::Future::Future() : type_(c10::NoneType::get()) {}
```
Comment (zdevito) [inline aten/src/ATen/core/ivalue.cpp]: Should not be part of this PR.

**Review by zdevito — COMMENTED** [review]
No issues beyond what @ezyang has mentioned.

File: aten/src/ATen/core/ivalue.cpp (hunk 6 above)
Comment (bwasti) [reply aten/src/ATen/core/ivalue.cpp]: yea, sorry -- I mentioned that in an above comment (I didn't use ghstack for this). will go away

**Comment by bwasti** [comment]
updated -- the windows build failure looks legit so I'm looking into that

File: aten/src/ATen/core/ivalue_inl.h [hunk 7]
```diff
@@ -522,6 +527,30 @@ c10::optional<T> generic_to(
...
+    typename... Args,
+    typename Indices = c10::guts::make_index_sequence<sizeof...(Args)>,
+    c10::guts::enable_if_t<
+        !c10::guts::disjunction<
+            std::is_lvalue_reference<Args>...,
+            c10::guts::negation<std::is_constructible<IValue, Args>>...>::value,
+        std::nullptr_t> = nullptr>
+std::tuple<Args...> generic_to(IValue ivalue, _fake_type<std::tuple<Args...>>) {
+  auto vals = ivalue.toTuple()->elements();
+  TORCH_INTERNAL_ASSERT(vals.size() == sizeof...(Args));
```
Comment (ezyang) [inline aten/src/ATen/core/ivalue_inl.h]: Hmm... an assert? The existing guidance in this file is a bit murky: it is true most of the type tests are done using `AT_ASSERT`, but that old mean actually means that we haven't appropriately applied the guidance from https://github.com/pytorch/pytorch/issues/20287

I think ivalue tests should be done with `TORCH_CHECK`. The reasoning is that IValue is part of the public C++ API, and a user may very well attempt to extract an IValue at the wrong type. IT would be wrong for them to report an error message to PyTorch in that case.

**Review by ezyang — APPROVED** [review]
yay!
